File: app/src/main/java/com/example/heroesandroid/heroes/clientserver/Connection.java
package com.example.heroesandroid.heroes.clientserver;

import android.util.Log;

import java.io.IOException;
import java.net.Socket;

public class Connection
{
    private  Socket  mSocket = null;
    private  String  mHost   = null;
    private  int     mPort   = 0;

    public static final String LOG_TAG = "SOCKET";

    public Connection() {}

    public Connection (final String host, final int port)
    {
        this.mHost = host;
        this.mPort = port;
    }

    // Метод открытия сокета
    public boolean openConnection() throws Exception
    {
        // Если сокет уже открыт, то он закрывается
        boolean result = false;
        closeConnection();
        try {
            // Создание сокета
            mSocket = new Socket(mHost, mPort);
            result = true;
        } catch (IOException e) {
            throw new Exception("Невозможно создать сокет: "
                    + e.getMessage());
        }
        return result;
    }
    /**
     * Метод закрытия сокета
     */
    public void closeConnection()
    {
        if (mSocket != null && !mSocket.isClosed()) {
            try {
                mSocket.close();
            } catch (IOException e) {
                Log.e(LOG_TAG, "Ошибка при закрытии сокета :"
                        + e.getMessage());
            } finally {
                mSocket = null;
            }
        }
        mSocket = null;
    }

    public Socket getmSocket() {
        return mSocket;
    }

    /**
     * Метод отправки данных
     */
    public void sendData(byte[] data) throws Exception {
        // Проверка открытия сокета
        if (mSocket == null || mSocket.isClosed()) {
            throw new Exception("Ошибка отправки данных. " +
                    "Сокет не создан или закрыт");
        }
        // Отправка данных
        try {
            mSocket.getOutputStream().write(data);
            mSocket.getOutputStream().flush();
        } catch (IOException e) {
            throw new Exception("Ошибка отправки данных : "
                    + e.getMessage());
        }
    }
    @Override
    protected void finalize() throws Throwable
    {
        super.finalize();
        closeConnection();
    }
}

File: app/src/main/java/com/example/heroesandroid/Connection.kt
package com.example.heroesandroid

import android.system.Os.socket
import android.util.Log
import java.io.*
import java.net.Socket


object Connection {
    private var mSocket: Socket? = null
    var mHost: String? = null
    var mPort = 0
    var inBuf: BufferedReader? = null
    var outBuf: BufferedWriter? = null

    val LOG_TAG = "SOCKET"

    // Метод открытия сокета
    @Throws(Exception::class)
    fun openConnection(): Boolean {
        // Если сокет уже открыт, то он закрывается
        var result = false
        closeConnection()
        try {
            // Создание сокета
            mSocket = Socket(mHost, mPort)
            inBuf = BufferedReader(InputStreamReader(mSocket!!.getInputStream()))
            outBuf = BufferedWriter(OutputStreamWriter(mSocket!!.getOutputStream()))
            result = true
        } catch (e: IOException) {
            Log.e("SOCKET ERROR",
                "Невозможно создать сокет: "
                        + e.message
                        + " " + mHost + " " + mPort
            )
        }
        return result
    }

    /**
     * Метод закрытия сокета
     */
    fun closeConnection() {
        if (mSocket != null && !mSocket!!.isClosed) {
            try {
                mSocket!!.close()
            } catch (e: IOException) {
                Log.e(
                    LOG_TAG, ("Ошибка при закрытии сокета :"
                            + e.message)
                )
            } finally {
                mSocket = null
            }
        }
        mSocket = null
        inBuf = null
        outBuf = null
    }

    fun isClosed(): Boolean {
        if (mSocket == null) return true
        return mSocket!!.isClosed
    }

}